// angular-shepherd/esm2015/lib/shepherd.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shepherd.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import Shepherd from 'shepherd.js';
import { elementIsHidden } from './utils/dom';
import { makeButton } from './utils/buttons';
import * as i0 from "@angular/core";
export class ShepherdService {
    constructor() {
        this.confirmCancel = false;
        this.confirmCancelMessage = null;
        this.defaultStepOptions = {};
        this.errorTitle = null;
        this.isActive = false;
        this.messageForUser = null;
        this.modal = false;
        this.requiredElements = [];
        this.steps = [];
        this.tourName = undefined;
        this.tourObject = null;
    }
    /**
     * Get the tour object and call back
     * @return {?}
     */
    back() {
        this.tourObject.back();
    }
    /**
     * Cancel the tour
     * @return {?}
     */
    cancel() {
        this.tourObject.cancel();
    }
    /**
     * Complete the tour
     * @return {?}
     */
    complete() {
        this.tourObject.complete();
    }
    /**
     * Hides the current step
     * @return {?}
     */
    hide() {
        this.tourObject.hide();
    }
    /**
     * Advance the tour to the next step
     * @return {?}
     */
    next() {
        this.tourObject.next();
    }
    /**
     * Show a specific step, by passing its id
     * @param {?} id The id of the step you want to show
     * @return {?}
     */
    show(id) {
        this.tourObject.show(id);
    }
    /**
     * Start the tour
     * @return {?}
     */
    start() {
        this.isActive = true;
        this.tourObject.start();
    }
    /**
     * This function is called when a tour is completed or cancelled to initiate cleanup.
     * @param {?} completeOrCancel 'complete' or 'cancel'
     * @return {?}
     */
    onTourFinish(completeOrCancel) {
        this.isActive = false;
    }
    /**
     * Take a set of steps and create a tour object based on the current configuration
     * @param {?} steps An array of steps
     * @return {?}
     */
    addSteps(steps) {
        this._initialize();
        /** @type {?} */
        const tour = this.tourObject;
        // Return nothing if there are no steps
        if (!steps || !Array.isArray(steps) || steps.length === 0) {
            return;
        }
        if (!this.requiredElementsPresent()) {
            tour.addStep({
                buttons: [{
                        text: 'Exit',
                        action: tour.cancel
                    }],
                id: 'error',
                title: this.errorTitle,
                text: [this.messageForUser]
            });
            return;
        }
        steps.forEach((/**
         * @param {?} step
         * @return {?}
         */
        (step) => {
            if (step.buttons) {
                step.buttons = step.buttons.map(makeButton.bind(this), this);
            }
            tour.addStep(step);
        }));
    }
    /**
     * Observes the array of requiredElements, which are the elements that must be present at the start of the tour,
     * and determines if they exist, and are visible, if either is false, it will stop the tour from executing.
     * @private
     * @return {?}
     */
    requiredElementsPresent() {
        /** @type {?} */
        let allElementsPresent = true;
        /* istanbul ignore next: also can't test this due to things attached to root blowing up tests */
        this.requiredElements.forEach((/**
         * @param {?} element
         * @return {?}
         */
        (element) => {
            /** @type {?} */
            const selectedElement = document.querySelector(element.selector);
            if (allElementsPresent && (!selectedElement || elementIsHidden(selectedElement))) {
                allElementsPresent = false;
                this.errorTitle = element.title;
                this.messageForUser = element.message;
            }
        }));
        return allElementsPresent;
    }
    /**
     * Initializes the tour, creates a new Shepherd.Tour. sets options, and binds events
     * @private
     * @return {?}
     */
    _initialize() {
        /** @type {?} */
        const tourObject = new Shepherd.Tour({
            confirmCancel: this.confirmCancel,
            confirmCancelMessage: this.confirmCancelMessage,
            defaultStepOptions: this.defaultStepOptions,
            tourName: this.tourName,
            useModalOverlay: this.modal
        });
        tourObject.on('complete', this.onTourFinish.bind(this, 'complete'));
        tourObject.on('cancel', this.onTourFinish.bind(this, 'cancel'));
        this.tourObject = tourObject;
    }
}
ShepherdService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ShepherdService.ctorParameters = () => [];
/** @nocollapse */ ShepherdService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ShepherdService_Factory() { return new ShepherdService(); }, token: ShepherdService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ShepherdService.prototype.confirmCancel;
    /** @type {?} */
    ShepherdService.prototype.confirmCancelMessage;
    /** @type {?} */
    ShepherdService.prototype.defaultStepOptions;
    /** @type {?} */
    ShepherdService.prototype.errorTitle;
    /** @type {?} */
    ShepherdService.prototype.isActive;
    /** @type {?} */
    ShepherdService.prototype.messageForUser;
    /** @type {?} */
    ShepherdService.prototype.modal;
    /** @type {?} */
    ShepherdService.prototype.requiredElements;
    /** @type {?} */
    ShepherdService.prototype.steps;
    /** @type {?} */
    ShepherdService.prototype.tourName;
    /** @type {?} */
    ShepherdService.prototype.tourObject;
}
